internal/config/client.go

```diff
@@ -2,37 +2,40 @@ package config
 
 import (
 	"context"
+	"k8s.io/apimachinery/pkg/runtime/schema"
 
 	authzv1 "k8s.io/api/authorization/v1"
 	"k8s.io/apimachinery/pkg/api/meta"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/client-go/rest"
 	"sigs.k8s.io/controller-runtime/pkg/cache"
 	"sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/controller-runtime/pkg/cluster"
 )
 
 // newCachingClient is an alternative implementation of controller-runtime's
 // default client for manager.Manager.
 // The only difference is that this implementation sets `CacheUnstructured` to `true` to
-// cache unstructured objects.
+// cache unstructured objects, as per commit d2c35505 in that repository.
 func newCachingClient(cache cache.Cache, config *rest.Config, options client.Options, uncachedObjects ...client.Object) (client.Client, error) {
-	c, err := client.New(config, options)
+	o := &options
+	o.Cache = &client.CacheOptions{
+		Reader:       cache,
+		DisableFor:   uncachedObjects,
+		Unstructured: true,
+	}
+	c, err := client.New(config, *o)
 	if err != nil {
 		return nil, err
 	}
-
-	return client.NewDelegatingClient(client.NewDelegatingClientInput{
-		CacheReader:       cache,
-		Client:            c,
-		UncachedObjects:   uncachedObjects,
-		CacheUnstructured: true,
-	})
+	return c, nil
 }
 
-func NewClient(readOnly bool) cluster.NewClientFunc {
-	return func(cache cache.Cache, config *rest.Config, options client.Options, uncachedObjects ...client.Object) (client.Client, error) {
-		c, err := newCachingClient(cache, config, options, uncachedObjects...)
+func NewClient(readOnly bool) client.NewClientFunc {
+	return func(config *rest.Config, options client.Options) (client.Client, error) {
+		// FIXME: mgr.GetCache() gives us the cache, but this is called when creating
```

Comment:
Need help with this one. To me it looks like we need the cache that we get from the manager, while we are in the process of creating the manager.

I'm fairly sure the idea conveyed with my code changes in `newCachingClient()` is correct though, it's now possible to create a caching client by passing the right options to `client.New()`.

Reply:
Ok, I removed the constant `nil` parameter and the FIXME. The `uncachedObjects` argument will never be passed, but I guess that is ok.